netplay: Valve GNS network backend implementation

ManManson opened the pull request:
The PR introduces a second network backend implementation for Warzone 2100, based on the Valve GameNetworkingSockets library (https://github.com/ValveSoftware/GameNetworkingSockets). This is a cross-platform, UDP-based transport layer designed specifically for games, packed with some advanced features, which WZ makes (or will make) use of:
* Connection-based APIs (TCP-like)
* Reliable messaging
* Built-in encryption
* P2P capabilities

This changeset includes a brand-new `GNSConnectionProvider` implementation, along with a set of supporting classes to represent client connections, listen sockets, poll groups and connection addresses, as well, as the necessary plumbing on the WZ side.

`WzConnectionProvider` interface has been extended with the `processConnectionStateChanges()` method, which should be called at the beginning of each game loop iteration.
This allows GNS library to process connection state changes and allows WZ to react to these by calling WZ-specific callbacks for each affected connection (connection initiated/dropped/accepted).

Warzone 2100 builds without the GNS support by default.
One would need to set `-DENABLE_GNS_NETWORK_BACKEND=ON` cmake flag while configuring the solution to enable the GNS network backend support. All CI builds have this flag enabled.

As WZ uses libsodium for all crypto-related things, GNS is force-configured to use libsodium, as well. The library uses encryption by default, so this doesn't need to be explicitly enabled by the users of the library.

**NOTE 1 (for hosters)**: Even when the WZ has been built with GNS support, it's still needed to be enabled explicitly in runtime to be able host games via GNS. This can be achieved by adding the `--host-connection-provider=gns` command line argument. WZ will then report the connection option used in the chat widget in hosting room.

**NOTE 2 (for joiners)**: The `--host-connection-provider` argument is not needed for those, who only needs to join a game, but doesn't want to host games themselves. ~~WZ will always try to connect via GNS first and then will fall back to the legacy TCP option.~~ **EDIT:** WZ will always try to connect via TCP first and if it's not available, will fall back to GNS connection option.

**NOTE 3 (lobby availability)**: At the moment, games hosted via GNS will not be listed in the lobby games list, since it requires additional support on the lobby server side (lobby server isn't able to talk directly to the host via the advertised connection option, because it doesn't currently understand the GNS's binary protocol).

past-due commented on lib/netplay/netplay.cpp:
```diff
@@ -2891,6 +2891,9 @@ static void NETcheckPlayers()
 // We should not block here.
 bool NETrecvNet(NETQUEUE *queue, uint8_t *type)
 {
+	ASSERT_OR_RETURN(false, activeConnProvider != nullptr, "No active connection provider!");
+	activeConnProvider->processConnectionStateChanges();
+
```
The `ASSERT_OR_RETURN` is problematic, as this condition is expected in at least challenge / skirmish lobbies and games, as well as when playing back a replay. Perhaps just a conditional?

```suggestion
	if (activeConnProvider)
	{
		activeConnProvider->processConnectionStateChanges();
	}

```

Also, could we safely move this below the `if (!NetPlay.bComms)` check? (Note: replays of multiplayer games would restore `NetPlay.bComms = true`, and `activeConnProvider` would be expected to be null.)

ManManson replied:
Sure, you're right. I'll fix that.

KJeff01 reviewed (commented):
Just writing to say I put in the effort to look at the code. Nothing super obvious stood out. Aside from that, there are things I noted in the Discord section about leave events during tests (also, there are longer wait times regarding incorrect password inputs while attempting to join--pastdue said that isn't normal).

ManManson commented:
I've implemented a lot of fixes related to error handling at various places (connection state change callbacks, pending writes manager error handling, additional state validation for client connections, connection options retries while joining a game).

Protocol selection preference for joiners has been changed from GNS-first to TCP-first, since GNS is still to be considered experimental and it won't be desirable to incur additional waiting time for those who still uses TCP to host and join games (which would be the majority of players at the time being).

Also, added adjustable connected timeout for client connections. This currently works the following way:

1. Initial connection timeout is 15 seconds (as it was before)
2. Connected timeout is set to 10 seconds upon connection initialization. This will be in effect while players are waiting in the lobby game room. In this case we would like to have smaller timeout so that in case someone is stuck or suddenly drops, the player slot won't be occupied for a long period of time.
3. Upon transitioning from the lobby wait to the main game loop, connected timeout value is increased to 60 seconds for all client connections, so that the automatic lag-kick routine will be able to take care of stuck connections.

So, everyone is encouraged to re-test the changeset again. 😄

past-due commented on src/main.cpp:
```diff
@@ -1043,6 +1043,9 @@ static void startGameLoop()
 			}
+			// Adjust connected timeouts for clients (which is typically larger than the value of timeout when
+			// waiting in a lobby room).
+			NETadjustConnectedTimeoutForClients();
```
We probably want this to be at the beginning of this function, before the call to `levLoadData` (which can currently block the main thread for an indeterminate - and potentially lengthy - amount of time).

ManManson replied:
Thanks for spotting this. Fixed.

ManManson commented:
Rebased the changeset on top of the latest master, updated GNS submodule (fixes issues with host unavailability on Windows) and improved error handling of dropped client connections.

past-due commented on src/screens/joiningscreen.cpp:
```diff
@@ -1292,7 +1309,11 @@ bool WzJoiningGameScreen_HandlerRoot::joiningSocketNETsend()
 		return false;
 	}
-	client_transient_socket->flush(nullptr);  // Make sure the message was completely sent.
+	// Make sure the message was completely sent.
+	if (!client_transient_socket->flush(nullptr).has_value())
+	{
+		return false;
+	}
 	ASSERT(queue->numMessagesForNet() == 0, "Queue not empty (%u messages remaining).", queue->numMessagesForNet());
```
```suggestion
	ASSERT(queue->numMessagesForNet() == 0, "Queue not empty (%u messages remaining).", queue->numMessagesForNet());
	// Make sure the message was completely sent.
	if (!client_transient_socket->flush(nullptr).has_value())
	{
		return false;
	}
```
Let's move the `ASSERT` before the call to `flush()`

past-due commented:
Huge thanks to @ManManson for the fantastic work refactoring netplay, adding GNS support (and tracking down & fixing bugs in GNS itself), and working through feedback. 🥳
